sda-commons-shared-asyncapi/src/main/java/org/sdase/commons/shared/asyncapi/internal/JsonSchemaEmbedder.java

```diff
@@ -0,0 +1,138 @@
+    } else if (current.isObject()) {
+      List<Entry<String, JsonNode>> fields = new ArrayList<>();
+      current.fields().forEachRemaining(fields::add);
+
+      fields.forEach(
+          e -> {
+            if (REF.equals(e.getKey()) && e.getValue().isTextual()) {
+              JsonReference ref = JsonReference.parse(e.getValue().asText());
+
+              embedReference(
+                  target, root, (ObjectNode) current, resolvedReferences, ref, shouldEmbedInternal);
+            } else {
+              resolveInPlace(target, root, e.getValue(), resolvedReferences, shouldEmbedInternal);
+            }
+          });
+    }
+  }
+
+  private void embedReference(
+      JsonNode target,
+      JsonNode root,
+      ObjectNode currentObject,
+      Map<String, JsonNode> resolvedReferences,
+      JsonReference ref,
+      boolean shouldEmbedInternal) {
+    JsonNode refDocument =
+        ref.isExternal()
+            ? resolvedReferences.computeIfAbsent(ref.url, jsonSchemaResolver::resolve)
+            : root;
+
+    // Check if external reference
+    if (refDocument != null && (ref.isExternal() || shouldEmbedInternal)) {
+      JsonPointer refNamePointer =
+          ref.pointer.last() == null
+              ? ref.pointer.append(JsonPointer.compile("/" + replaceSpecialCharacters(ref.url)))
+              : ref.pointer.last();
+      String refName = refNamePointer.getMatchingProperty();
+
+      JsonNode embedNode = ensureParentNode(target, embedPointer);
+      ObjectNode embedObject = (ObjectNode) embedNode;
+
+      // Not yet embedded?
+      if (!embedObject.has(refName)) {
+        JsonNode refNode = refDocument.at(ref.pointer).deepCopy();
+
+        if (ref.pointer.last() == null) {
+          // In case the root object is embedded, we have to remove the other unrelated top level
+          // fields like $schema or definitions
+          ObjectNode refObject = (ObjectNode) refNode;
+          IGNORE_ON_EMBED.forEach(refObject::remove);
+        }
+
+        // Make sure that all internal references are also recursively embedded
+        resolveInPlace(target, refDocument, refNode, resolvedReferences, true);
+
+        embedObject.set(refName, refNode);
+      }
+
+      // Rewrite reference to embedded location
+      currentObject.put(REF, new JsonReference(embedPointer.append(refNamePointer)).toString());
+    }
+  }
+
+  private JsonNode ensureParentNode(JsonNode target, JsonPointer pointer) {
+    JsonNode node = target.at(pointer);
+
+    if (node.isMissingNode()) {
+      ObjectNode parentNode = (ObjectNode) ensureParentNode(target, pointer.head());
+      parentNode.set(pointer.last().getMatchingProperty(), JsonNodeFactory.instance.objectNode());
+      node = target.at(embedPointer);
+    }
+
+    return node;
+  }
+
+  private String replaceSpecialCharacters(String value) {
+    return value.replaceAll("[^a-zA-Z0-9]", "");
```

Comment:
too strict for URLs? is it important to represent the URL properly?

Reply:
The idea is to generate a temporary and unique class name, I don't really care about how it looks, just that it is safe. 
Url will be a relative path most of the time, so I wouldn't consider urls here right now.